control-plane/agents/core/src/volume/tests.rs

```diff
@@ -0,0 +1,459 @@
+    let spec = nexus_spec(&nexus).await.unwrap();
+    assert!(spec.operation.unwrap().result.is_some());
+
+    // thaw etcd allowing the worker thread to sync the "dirty" spec
+    cluster.composer().thaw("etcd").await.unwrap();
+
+    // wait for the reconciler to do its thing
+    tokio::time::delay_for(reconcile_period * 2).await;
+
+    // and now we're in sync and the pending operation is no more
+    let spec = nexus_spec(&nexus).await.unwrap();
+    assert!(spec.operation.is_none());
+    assert_eq!(spec.share, Protocol::Nvmf);
+
+    share
+        .request_ext(bus_timeout_opts())
+        .await
+        .expect_err("already shared via nvmf");
+}
+
+#[actix_rt::test]
+async fn nexus_add_child_transaction() {
+    let cluster = ClusterBuilder::builder()
+        .with_rest(false)
+        .with_pools(1)
+        .with_agents(vec!["core"])
+        .with_node_timeouts(Duration::from_millis(250), Duration::from_millis(500))
+        .with_bus_timeouts(bus_timeout_opts())
+        .build()
+        .await
+        .unwrap();
+    let mayastor = cluster.node(0);
+
+    let nodes = GetNodes {}.request().await.unwrap();
+    tracing::info!("Nodes: {:?}", nodes);
+
+    let child2 = "malloc:///ch2?size_mb=12";
+    let nexus = CreateNexus {
+        node: mayastor.clone(),
+        uuid: "f086f12c-1728-449e-be32-9415051090d6".into(),
+        size: 5242880,
+        children: vec!["malloc:///ch1?size_mb=12".into()],
+        ..Default::default()
+    }
+    .request()
+    .await
+    .unwrap();
+    let add_child = AddNexusChild {
+        node: mayastor.clone(),
+        nexus: nexus.uuid.clone(),
+        uri: child2.into(),
+        auto_rebuild: true,
+    };
+    let rm_child = RemoveNexusChild {
+        node: mayastor.clone(),
+        nexus: nexus.uuid.clone(),
+        uri: child2.into(),
+    };
+
+    async fn check_add_operation(nexus: &Nexus, children: usize) {
+        // operation in progress
+        assert!(nexus_spec(&nexus).await.unwrap().operation.is_some());
+        tokio::time::delay_for(std::time::Duration::from_millis(500)).await;
+        // operation is complete
+        assert!(nexus_spec(&nexus).await.unwrap().operation.is_none());
+        assert_eq!(nexus_spec(&nexus).await.unwrap().children.len(), children);
+    }
+
+    // pause mayastor
+    cluster.composer().pause(mayastor.as_str()).await.unwrap();
+
+    add_child
+        .request_ext(bus_timeout_opts())
+        .await
+        .expect_err("mayastor is down");
+
+    check_add_operation(&nexus, 1).await;
+
+    // unpause mayastor
+    cluster.composer().thaw(mayastor.as_str()).await.unwrap();
+
+    // now it should be shared successfully
+    let uri = add_child.request().await.unwrap();
+    println!("Share uri: {:?}", uri);
+
+    cluster.composer().pause(mayastor.as_str()).await.unwrap();
+
+    rm_child
+        .request_ext(bus_timeout_opts())
+        .await
+        .expect_err("mayastor down");
+
+    check_add_operation(&nexus, 2).await;
```

Comment:
Maybe this function should be renamed as we are removing a child at this point.

Reply:
Done thanks. Also modified the other test add_child...store to include the removal case so both are tested